### PR TITLE
Provide `gitlab_pages_domain` resource

### LINKED ISSUE #498: Feature request - Expose DNS records used in Pages configuration to rest of Terraform
Currently configuring Pages with a custom domain requires a manual intervention.

### Terraform Version
```
Terraform v0.13.5
```

### Affected Resource(s)
If applicable: 
- gitlab_project
- gitlab_project_pages (if it will exist in next release)

### Actual Behavior
Jane is working on a static site with GitLab Pages and already has a DNS provider that supports Terraform. She has to manually copy and paste DNS records for her custom domain from the GitLab console into her Terraform code, which is **troublesome and error prone**.

### Expected Behavior
Jane wishes to get the following from the GitLab provider:
- CNAME DNS record that directs traffic from her domain to Pages;
- TXT DNS record used for verification;

by ex. using a **data source**. That way she could **pass those records as variables** to the provider of her DNS company. This scales much better and is much less error prone than semi-automatic configuration.

### Important Factoids
GitLab provider version:
- 3.1.0
- 3.2.0

### References

- Similar case, inproving functionality of Pages for Terraform users: https://github.com/gitlabhq/terraform-provider-gitlab/issues/471
- Documentation: https://docs.gitlab.com/ee/user/project/pages/custom_domains_ssl_tls_certification/index.html#4-verify-the-domains-ownership

### PR DESCRIPTION
**Pull request by nagyv**
## Description

Create a `gitlab_pages_domain` resource

Closes #498

Blocked by https://github.com/xanzy/go-gitlab/pull/1452

### PR Checklist Acknowledgement

- [x] I acknowledge that all of the following items are true, where applicable:
  - Resource attributes match 1:1 the names and structure of the API resource in [the GitLab API documentation](https://docs.gitlab.com/ee/api/). ✅ 
  - [Examples](https://github.com/gitlabhq/terraform-provider-gitlab/tree/main/examples) are updated with:
    - A \*.tf file for the resource/s with at least one usage example
    - A \*.sh file for the resource/s with an import example (if applicable)
  - New resources have at minimum a basic test with three steps:
    - Create the resource ✅ 
    - Update the attributes ✅ 
    - Import the resource
  - No new `//lintignore` comments were copied from existing code. (Linter rules are meant to be enforced on new code.) ✅

### CODE REVIEW

**Comment by nagyv** [comment]
I expect the tests to fail for now. I run into two issues with the tests:

1. Running `make testacc RUN=TestAccGitlabPagesDomain_basic` returns error 404
2. Running `make testacc RUN=TestAccGitlabPagesDomain_basic GITLAB_TOKEN=glpat-xxx GITLAB_BASE_URL=https://gitlab.com` returns error 400. 

The latter led to https://github.com/xanzy/go-gitlab/pull/1452

Why was there a different error in GitPod? Are GitLab pages enabled in GitPod?

**Comment by PatrickRice-KSC** [comment]
Hey @nagyv , yeah, pages isn't configured in the docker-compose/makefile yet. We could potentially add a new wildcard cert to the ones generated by openssl (should probably wait until we merge #1023 so we can build on top of the work shock is doing), which seems to be a pre-req for enabling pages based on: https://docs.gitlab.com/ee/administration/pages/index.html#configuration

image

**Comment by timofurrer** [comment]
FYI @PatrickRice-KSC @nagyv I've merged #1023 🎉

File: internal/provider/resource_gitlab_pages_domain.go
Comment (timofurrer) [inline internal/provider/resource_gitlab_pages_domain.go:16]:
```suggestion
		Description: `The ` + "`gitlab_pages_domain`" + ` resource allows to manage the lifecycle of a custom Pages domain including its TLS certificates.
```

File: internal/provider/resource_gitlab_pages_domain.go
Comment (timofurrer) [inline internal/provider/resource_gitlab_pages_domain.go:36]:
```suggestion
				Description: "The ID or [URL-encoded path of the project](https://docs.gitlab.com/ee/api/index.html#namespaced-path-encoding) owned by the authenticated user.",
```

File: internal/provider/resource_gitlab_pages_domain.go
Comment (timofurrer) [inline internal/provider/resource_gitlab_pages_domain.go:46]: NIT: we usually do not specify the defaults

```suggestion
```

File: internal/provider/resource_gitlab_pages_domain.go
Comment (timofurrer) [inline internal/provider/resource_gitlab_pages_domain.go:53]: NIT: we usually do not specify the defaults

```suggestion
```

File: internal/provider/resource_gitlab_pages_domain.go
Comment (timofurrer) [inline internal/provider/resource_gitlab_pages_domain.go:60]: NIT: we usually do not specify the defaults

```suggestion
```

File: internal/provider/resource_gitlab_pages_domain.go
Comment (timofurrer) [inline internal/provider/resource_gitlab_pages_domain.go:75]: What about the `subject` ?

File: internal/provider/resource_gitlab_pages_domain.go
Comment (timofurrer) [inline internal/provider/resource_gitlab_pages_domain.go:101]: Are these undocumented fields?

File: internal/provider/resource_gitlab_pages_domain.go
Comment (timofurrer) [inline internal/provider/resource_gitlab_pages_domain.go:108]: NIT:

```suggestion
	projectID := d.Get("project").(string)
```

File: internal/provider/resource_gitlab_pages_domain.go
Comment (timofurrer) [inline internal/provider/resource_gitlab_pages_domain.go:123]: NIT:

```suggestion
	_, _, err := client.PagesDomains.CreatePagesDomain(projectID, options, gitlab.WithContext(ctx))
```

File: internal/provider/resource_gitlab_pages_domain.go
Comment (timofurrer) [inline internal/provider/resource_gitlab_pages_domain.go:128]: NIT:

```suggestion
	d.SetId(buildTwoPartID(&projectID, &domain))
```

File: internal/provider/resource_gitlab_pages_domain.go
Comment (timofurrer) [inline internal/provider/resource_gitlab_pages_domain.go:119]: NIT: to indicate that we only use these attributes for the GitLab options I think it makes sense to not have variables for them ...

```suggestion
	domain := d.Get("domain").(string)
	auto_ssl_enabled := d.Get("auto_ssl_enabled").(bool)
	certificate := d.Get("certificate").(string)
	key := d.Get("key").(string)

	options := &gitlab.CreatePagesDomainOptions{
		Domain:         gitlab.String(d.Get("domain").(string)),
		AutoSslEnabled: gitlab.Bool(d.Get("auto_ssl_enabled").(bool))
		Certificate:    gitlab.String(d.Get("certificate").(string)),
		Key:            gitlab.String(d.Get("key").(string)),
	}
```

Please verify the correct behavior during creation if these optional attributes are not given and that the GitLab API doesn't receive them ....

File: internal/provider/resource_gitlab_pages_domain.go
Comment (timofurrer) [inline internal/provider/resource_gitlab_pages_domain.go:135]: It's a common practise to use the `id` here as well, like you did for the `Read()`

File: internal/provider/resource_gitlab_pages_domain.go
Comment (timofurrer) [inline internal/provider/resource_gitlab_pages_domain.go:141]: I'd only update the fields for which the attribute in the tf config actually changed.

You can use e.g. `d.HasChange("auto_ssl_enabled")` to check that and add it to the `options`.

File: internal/provider/resource_gitlab_pages_domain.go
Comment (timofurrer) [inline internal/provider/resource_gitlab_pages_domain.go:207]: Can `pagesDomain.Certificate` be `nil` ?

File: internal/provider/resource_gitlab_pages_domain_test.go
Comment (timofurrer) [inline internal/provider/resource_gitlab_pages_domain_test.go:16]: NIT: early exit asap.

```suggestion
	testAccCheck(t)
	
	var pagesDomain gitlab.PagesDomain
	rInt := acctest.RandInt()
```

File: internal/provider/resource_gitlab_pages_domain_test.go
Comment (timofurrer) [inline internal/provider/resource_gitlab_pages_domain_test.go:24]: Let's add another test case to only supply the required data - there may be some issues in the current implementation when `pagesDomain.Certificate` is `nil` or even earlier when `go-gitlab` fails, because `nil` is passed to `gitlab.String()` - these are just suspicions though and everything might be okay as-is 👯

File: internal/provider/resource_gitlab_pages_domain_test.go
Comment (timofurrer) [inline internal/provider/resource_gitlab_pages_domain_test.go:29]: Given the subsequent `Verify import` test step both of these checks are already done, because:

1. `import` does a `Read()` and therefore the Pages domain actually has to exist
2. it would result in a non-empty plan in case the `auto_ssl_enabled` is not properly set ...

File: internal/provider/resource_gitlab_pages_domain_test.go
Comment (timofurrer) [inline internal/provider/resource_gitlab_pages_domain_test.go:123]: In case they don't grow after the feedback implementation I think we can inline them for better reason about the test cases.

**Review by timofurrer — CHANGES_REQUESTED** [review]
A few minor feedbacks and nit picks no linter is able to catch yet ...

Looks solid otherwise 🎉

File: internal/provider/resource_gitlab_pages_domain_test.go
Comment (nagyv) [reply internal/provider/resource_gitlab_pages_domain_test.go]: Sorry, I don't understand your suggestion. :)

File: internal/provider/resource_gitlab_pages_domain_test.go
Comment (nagyv) [reply internal/provider/resource_gitlab_pages_domain_test.go:30]: What is your recommendation? I would prefer to have a "create" test case, beside the import.

File: internal/provider/resource_gitlab_pages_domain.go
Comment (nagyv) [reply internal/provider/resource_gitlab_pages_domain.go]: Even if we remove the default, I'd keep the `ForceNew`. The docs were not clear what the behaviour is if someone changes the `auto_ssl_enabled` flag. Re-creating the domain seems to be a safe bet.

File: internal/provider/resource_gitlab_pages_domain.go
Comment (nagyv) [reply internal/provider/resource_gitlab_pages_domain.go]: I'd leave the ForceNew here too. I might be wrong, but the `certificate_data` is a computed property based on the `certificate`. The plan might be misleading without ForceNew, as `certificate_data` will be known only after the apply step. I guess ForceNew might achieve this.

File: internal/provider/resource_gitlab_pages_domain.go
Comment (nagyv) [reply internal/provider/resource_gitlab_pages_domain.go]: I added ForceNew for the same reason as with the certificate. Let me know if my reasoning is wrong. I likely know less about terraform internals than you do.

File: internal/provider/resource_gitlab_pages_domain.go
Comment (nagyv) [reply internal/provider/resource_gitlab_pages_domain.go:72]: https://github.com/xanzy/go-gitlab/blob/master/pages_domains.go#L44 returns only the expired and expiration keys.

File: internal/provider/resource_gitlab_pages_domain.go
Comment (nagyv) [reply internal/provider/resource_gitlab_pages_domain.go:98]: Yes. https://github.com/gitlabhq/terraform-provider-gitlab/issues/498#issuecomment-734532593

File: internal/provider/resource_gitlab_pages_domain.go
Comment (nagyv) [reply internal/provider/resource_gitlab_pages_domain.go:116]: Do you mean separate testcases? The current test does not specify any of the optional attributes.

File: internal/provider/resource_gitlab_pages_domain.go
Comment (nagyv) [reply internal/provider/resource_gitlab_pages_domain.go:204]: No. I tried that first, but it's always initialized with `Expired: false, Expiration: nil`

File: internal/provider/resource_gitlab_pages_domain_test.go
Comment (nagyv) [reply internal/provider/resource_gitlab_pages_domain_test.go:25]: I removed the `Default` values and changed to testcase to include only the required ones.

File: internal/provider/resource_gitlab_pages_domain.go
Comment (timofurrer) [reply internal/provider/resource_gitlab_pages_domain.go]: hmm, I would assume that if it can be updated, it safe from the API side, now? 

But then we should set `ForceNew: true` anyways, and not `ForceNew: false`, right?

File: internal/provider/resource_gitlab_pages_domain.go
Comment (timofurrer) [reply internal/provider/resource_gitlab_pages_domain.go]: I don't get that 😅  ... I mean `ForceNew: false` is the default anyways, I'm just saying we should remove from the schema - so that it's non-explicit.

File: internal/provider/resource_gitlab_pages_domain.go
Comment (timofurrer) [reply internal/provider/resource_gitlab_pages_domain.go:72]: ... which is somewhat misaligned with the upstream API ... 

In case I have some time tomorrow I'll update the GitLab docs and `go-gitlab`. Otherwise, I think we go with what's there :)

File: internal/provider/resource_gitlab_pages_domain.go
Comment (timofurrer) [reply internal/provider/resource_gitlab_pages_domain.go:116]: True, noticed afterwards - let's make sure that both scenarios are tested: creation without optionals set, creation with optionals set ...

File: internal/provider/resource_gitlab_pages_domain_test.go
Comment (timofurrer) [reply internal/provider/resource_gitlab_pages_domain_test.go]: oh sorry, this might help: https://github.com/gitlabhq/terraform-provider-gitlab/issues/1008#:~:text=Prefer%20writing%20config%20in%2Dline%20to%20wrapping%20it%20in%20a%20function.

**Comment by timofurrer** [comment]
@nagyv btw. I think something got messed up during the rebase ....

File: internal/provider/resource_gitlab_pages_domain_test.go
Comment (timofurrer) [inline internal/provider/resource_gitlab_pages_domain_test.go:38]:
```suggestion
```

There is not `token`

File: internal/provider/resource_gitlab_pages_domain_test.go
Comment (timofurrer) [reply internal/provider/resource_gitlab_pages_domain_test.go:25]: As per my other comment, let's have multiple test cases: testing w/ and w/o optionals set.

File: internal/provider/resource_gitlab_pages_domain_test.go
Comment (timofurrer) [reply internal/provider/resource_gitlab_pages_domain_test.go:30]: Maybe this helps: [Use ImportStateVerify TestSteps as a replacement for writing custom "CheckExists" and "CheckAttributes" check funcs.](https://github.com/gitlabhq/terraform-provider-gitlab/issues/1008#:~:text=and%20data%20sources.-,Use%20ImportStateVerify%20TestSteps%20as%20a%20replacement%20for%20writing%20custom%20%22CheckExists%22%20and%20%22CheckAttributes%22%20check%20funcs.,-Undesirable%3A)

Also see the best practice here: https://github.com/bflad/tfproviderlint/tree/main/passes/AT002#at002

File: internal/provider/resource_gitlab_pages_domain.go
Comment (nagyv) [inline internal/provider/resource_gitlab_pages_domain.go:57]: Remove default

```suggestion
```

File: internal/provider/resource_gitlab_pages_domain.go
Comment (nagyv) [inline internal/provider/resource_gitlab_pages_domain.go:51]: Remove default

```suggestion
```

File: internal/provider/resource_gitlab_pages_domain.go
Comment (nagyv) [inline internal/provider/resource_gitlab_pages_domain.go:45]: Remove default

```suggestion
```

**Pull request closed without merging**
